src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/HashProviderDispenser.OSX.cs

```diff
@@ -0,0 +1,233 @@
+                case HashAlgorithmNames.SHA384:
+                    return new AppleDigestProvider(Interop.AppleCrypto.PAL_HashAlgorithm.Sha384);
+                case HashAlgorithmNames.SHA512:
+                    return new AppleDigestProvider(Interop.AppleCrypto.PAL_HashAlgorithm.Sha512);
+            }
+
+            throw new PlatformNotSupportedException();
+        }
+
+        public static HashProvider CreateMacProvider(string hashAlgorithmId, byte[] key)
+        {
+            switch (hashAlgorithmId)
+            {
+                case HashAlgorithmNames.MD5:
+                    return new AppleHmacProvider(Interop.AppleCrypto.PAL_HashAlgorithm.Md5, key);
+                case HashAlgorithmNames.SHA1:
+                    return new AppleHmacProvider(Interop.AppleCrypto.PAL_HashAlgorithm.Sha1, key);
+                case HashAlgorithmNames.SHA256:
+                    return new AppleHmacProvider(Interop.AppleCrypto.PAL_HashAlgorithm.Sha256, key);
+                case HashAlgorithmNames.SHA384:
+                    return new AppleHmacProvider(Interop.AppleCrypto.PAL_HashAlgorithm.Sha384, key);
+                case HashAlgorithmNames.SHA512:
+                    return new AppleHmacProvider(Interop.AppleCrypto.PAL_HashAlgorithm.Sha512, key);
+            }
+
+            throw new PlatformNotSupportedException();
+        }
+
+        // -----------------------------
+        // ---- PAL layer ends here ----
+        // -----------------------------
+
+        private class AppleHmacProvider : HashProvider
+        {
+            private readonly byte[] _key;
+            private readonly SafeHmacHandle _ctx;
+
+            private bool _running;
+
+            public override int HashSizeInBytes { get; }
+
+            internal AppleHmacProvider(Interop.AppleCrypto.PAL_HashAlgorithm algorithm, byte[] key)
+            {
+                _key = key.CloneByteArray();
+                int hashSizeInBytes = 0;
+                _ctx = Interop.AppleCrypto.HmacCreate(algorithm, ref hashSizeInBytes);
+
+                if (hashSizeInBytes < 0)
+                {
+                    _ctx.Dispose();
+                    throw new PlatformNotSupportedException(
+                        SR.Format(
+                            SR.Cryptography_UnknownHashAlgorithm,
+                            Enum.GetName(typeof(Interop.AppleCrypto.PAL_HashAlgorithm), algorithm)));
+                }
+
+                if (_ctx.IsInvalid)
+                {
+                    throw new CryptographicException();
+                }
+
+                HashSizeInBytes = hashSizeInBytes;
+            }
+
+            public override unsafe void AppendHashDataCore(byte[] data, int offset, int count)
+            {
+                if (!_running)
+                {
+                    SetKey();
+                }
+
+                int ret;
+
+                fixed (byte* pData = data)
```

Comment:
Okay.  It looks like we do have an explicit test for this (phew!). The MD5 tests, from RFC 1321 A.5, start off with the empty input case.  For that we send the empty array; which gets turned into null by the `fixed` statement, but apparently all of the providers are willing to accept cbData==0 as success even when pbData is null.